rcmgr.go

```diff
@@ -0,0 +1,441 @@
+package rcmgr
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+)
+
+type ResourceManager struct {
+	limits Limiter
+
+	system    *SystemScope
+	transient *TransientScope
+
+	mx    sync.Mutex
+	svc   map[string]*ServiceScope
+	proto map[protocol.ID]*ProtocolScope
+	peer  map[peer.ID]*PeerScope
+
+	cancelCtx context.Context
+	cancel    func()
+	wg        sync.WaitGroup
+}
+
+var _ network.ResourceManager = (*ResourceManager)(nil)
+
+type SystemScope struct {
+	*ResourceScope
+}
+
+var _ network.ResourceScope = (*SystemScope)(nil)
+
+type TransientScope struct {
+	*ResourceScope
+
+	system *SystemScope
+}
+
+var _ network.ResourceScope = (*TransientScope)(nil)
+
+type ServiceScope struct {
```

Comment:
This doesn't need to be exported, does it?
Same for all the other scopes defined here.